lucene/core/src/java/org/apache/lucene/util/fst/FST.java

```diff
@@ -865,20 +939,32 @@ public int readNextArcLabel(Arc<T> arc, BytesReader in) throws IOException {
       } else {
         in.setPosition(pos);
       }
+      // skip flags
+      in.readByte();
     } else {
       if (arc.bytesPerArc != 0) {
         //System.out.println("    nextArc real array");
-        // arcs are at fixed entries
-        in.setPosition(arc.posArcsStart);
-        in.skipBytes((1+arc.arcIdx)*arc.bytesPerArc);
+        // arcs are in an array
+        if (arc.arcIdx >= 0) {
+          in.setPosition(arc.posArcsStart);
+          // point at next arc, -1 to skip flags
+          in.skipBytes((1 + arc.arcIdx) * arc.bytesPerArc + 1);
+        } else {
+          in.setPosition(arc.nextArc);
+          byte flags = in.readByte();
+          // skip missing arcs
+          while (flag(flags, BIT_MISSING_ARC)) {
```

Comment:
It would be nice (separate issue) if we could track counters of how many "arc scans" (typically done by `seekCeil` from Lucene's terms dict) versus "direct lookups" are done, in a typical use case ... but that's a future issue.

Reply:
great idea - perhaps if we add some Arc abstractions that would provide a nice hook for such tracking/counting